R/read_vpts.R

```diff
@@ -1,241 +1,144 @@
+    csv = read_vpts_csv(files),
+    gz = read_vpts_csv(files),
+    h5 = read_vpts_hdf5(files)
   )
-  # read the data
-
-  if(sep!=""){
-    # for parsing new csv format
-    data <- utils::read.table(file = file, header = TRUE, sep = sep)
-    radar <- unique(data$radar)
-    if(length(radar)>1){
-      stop("file contains data for multiple radars")
-    }
-    data$radar <- NULL
-    data$datetime <- readr::parse_datetime(data$datetime)
-    data$gap <- as.logical(data$gap)
-  } else{
-    # for parsing legacy vol2bird text output
-    data <- utils::read.table(file = file, header = FALSE, sep = sep)
-    if (ncol(data) == 22) {
-      colnames(data) <- header.names.long
-    } else {
-      colnames(data) <- header.names.short
-    }
-
-    # convert Time into a POSIXct date-time
-    data$datetime <- as.POSIXct(paste(data$Date, sprintf("%04d", data$Time),
-                                      sep = ""
-    ),
-    format = "%Y%m%d%H%M",
-    tz = "UTC"
-    )
-  }
 
-  # add profile_index to identify consecutive profiles
-  data$new_profile_starts <- c(T, (data$height[-1] - data$height[-length(data$height)]) < 0)
-  data$profile_index <- NA
-  profile_index <- NULL # define profile_index to suppress devtools::check warning in next line
-  data[which(data$new_profile_starts), "profile_index"] <- 1:length(which(data$new_profile_starts))
-  data <- tidyr::fill(data, profile_index)
-
-  data$new_profile_starts <- NULL
-  data$Date <- NULL
-  data$Time <- NULL
-
-  # sort
-  data <- data[with(data, order(datetime, profile_index, height)), ]
-
-  # split into profiles
-  data <- split(data, data$profile_index)
-  names(data) <- NULL
-
-  # verify that profiles can be flattened
-  datadim <- sapply(1:length(data), function(x) dim(data[[x]]))
-
-  if (length(unique(datadim[1, ])) > 1) {
-    mostFrequent <- sort(table(datadim[1, ]), decreasing = TRUE)[1]
-    if (mostFrequent <= 1) {
-      stop("Profiles are of unequal altitudinal dimensions, unable to merge")
-    }
-    mostFrequentNBins <- as.integer(names(mostFrequent))
-    warning(paste(
-      "Profiles are of unequal altitudinal dimensions or",
-      "contain duplicates. Discarding", length(data) - mostFrequent,
-      "of", length(data), "profiles, restricting to",
-      mostFrequentNBins, "altitude bins."
-    ))
-    data <- data[datadim[1, ] == mostFrequentNBins]
-  }
+  data
+}
 
-  # strip the datetime field
-  datetime <- .POSIXct(sapply(
-    1:length(data),
-    function(x) {
-      data[[x]]$datetime[1]
-    }
-  ),
-  tz = "UTC"
+#' Read time series of vertical profiles (`vpts`) from VPTS CSV file(s)
+#'
+#' @inheritParams read_vpts
+#' @param df If `TRUE` returns data as data frame rather than `vpts`
+#'   object.
+#' @return `vpts` object.
+#' @noRd
+read_vpts_csv <- function(files, df = FALSE) {
+  # Create Frictionless Data Package
+  package <- frictionless::create_package()
+  schema <- "https://raw.githubusercontent.com/enram/vpts-csv/main/vpts-csv-table-schema.json"
```

Comment:
does this mean we can only read files when there is an internet connection? I think we want read_vpts to work also in absence of internet.

Reply:
Correct. If we want to support no connection, we need to store that schema locally in bioRad.